### PR TITLE
Refine CGContext's user<->device coordinate space boundary

### PR DESCRIPTION
**Pull request by DHowett-MSFT**
CGContext will now strive, wherever possible, to store its internal state in fully-device-transformed units. This transformation includes the device scale (flip+translate from CG to D2D) and the user scale.

To enable this, this pull request also adds the ability to specify that coordinate system a drawn geometry is in: paths will use the premultiplied values, and user-space geometry values will use userspace->device transformation.

This allows us to leverage the Direct2D effect stack without having to apply input and output transformations to the effects, and simplifies shadow rendering (we no longer transform for userspace in the commandlist and then device space when we draw it.)

---
This change is [Reviewable](https://reviewable.io/reviews/microsoft/winobjc/1324)

### CODE REVIEW

**Comment by DHowett-MSFT** [comment]
Refs #1268.

File: Frameworks/CoreGraphics/CGContext.mm [hunk 1]
```diff
@@ -354,67 +359,65 @@ void CGContextEndPage(CGContextRef context) {
...
 #pragma region Global State - Device Coordinate Queries
 /**
- @Status Stub
+ @Status Interoperable
 */
-CGRect CGContextConvertRectToDeviceSpace(CGContextRef context, CGRect rect) {
-    NOISY_RETURN_IF_NULL(context, StubReturn());
-    UNIMPLEMENTED();
-    return rect;
+CGRect CGContextConvertRectToUserSpace(CGContextRef context, CGRect rect) {
```
Comment (aballway) [inline Frameworks/CoreGraphics/CGContext.mm]: do we have tests for these?

File: Frameworks/CoreGraphics/CGContext.mm [hunk 2]
```diff
@@ -1429,20 +1443,35 @@ static HRESULT __CGContextRenderToCommandList(CGContextRef context, ID2D1Command
...
+            transform = CGContextGetUserSpaceToDeviceSpaceTransform(context);
+            break;
+        case _kCGCoordinateModeDeviceSpace:
+        default:
+            // do nothing; base transform is identity.
+            break;
+    }
+
+    if (additionalTransform) {
+        transform = CGAffineTransformConcat(*additionalTransform, transform);
```
Comment (rajsesh-msft) [inline Frameworks/CoreGraphics/CGContext.mm]: The order of the concat matters.  "additional transform" seems to indicate that it should be the second param?

**Comment by DHowett-MSFT** [comment]
@rajsesh-msft @aballway tests done been added!

File: Frameworks/CoreGraphics/CGContext.mm [hunk 3]
```diff
@@ -1527,7 +1552,7 @@ void CGContextStrokeRect(CGContextRef context, CGRect rect) {
     FAIL_FAST_IF_FAILED(factory->CreateRectangleGeometry(__CGRectToD2D_F(rect), &rectGeometry));
     FAIL_FAST_IF_FAILED(rectGeometry.As(&geometry));
 
-    FAIL_FAST_IF_FAILED(__CGContextDrawGeometry(context, geometry.Get(), kCGPathStroke));
+    FAIL_FAST_IF_FAILED(__CGContextDrawGeometry(context, _kCGCoordinateModeUserSpace, geometry.Get(), kCGPathStroke));
```
Comment (rajsesh-msft) [inline Frameworks/CoreGraphics/CGContext.mm]: Shouldn't these FAIL_FAST_IF_FAILED be RETURN_IF_FAILED() instead?  Or is your plan to do it once at some point?

File: tests/unittests/CoreGraphics/CGContextTests.mm [hunk 4]
```diff
@@ -262,3 +272,146 @@ void cgContextPathCompare(NSArray* expected, NSArray* result) {
     CGContextRelease(context);
     CGColorSpaceRelease(rgbColorSpace);
 }
+
+bool operator==(const CGPoint& lhs, const CGPoint& rhs) {
+    return ((std::abs(lhs.x - rhs.x) < 0.00001) && (std::abs(lhs.y - rhs.y) < 0.00001));
+}
+
+template <typename T>
+std::basic_ostream<T>& operator<<(std::basic_ostream<T>& os, const CGPoint& pt) {
```
Comment (rajsesh-msft) [inline tests/unittests/CoreGraphics/CGContextTests.mm]: 🛩

File: tests/unittests/CoreGraphics/CGContextTests.mm [hunk 5]
```diff
@@ -262,3 +272,146 @@ void cgContextPathCompare(NSArray* expected, NSArray* result) {
...
+    os << "[";
+    if (CGAffineTransformIsIdentity(transform)) {
+        os << "identity transform";
+    } else {
+        os << transform.a << " " << transform.b << " " << transform.c << " " << transform.d << " " << transform.tx << " " << transform.ty;
+    }
+    return os << "]";
+}
+
+class ContextCoordinateTest
```
Comment (rajsesh-msft) [inline tests/unittests/CoreGraphics/CGContextTests.mm]: You may want to keep this under win32 target for test portability for compat testing.

Comment (DHowett-MSFT) [inline tests/unittests/CoreGraphics/CGContextTests.mm]: Good call. I'll put that and the includes under TARGET

File: Frameworks/CoreGraphics/CGContext.mm (hunk 3 above)
Comment (DHowett-MSFT) [inline Frameworks/CoreGraphics/CGContext.mm]: Yep; once we're done here we can go through and adjust them vis-a-vis #1323

File: tests/unittests/CoreGraphics/CGContextTests.mm [hunk 6]
```diff
@@ -262,3 +274,148 @@ void cgContextPathCompare(NSArray* expected, NSArray* result) {
...
+        }
+    }
+};
+
+/* static */
+ComPtr<ID2D1RenderTarget> ContextCoordinateTest::renderTarget;
+
+TEST_P(ContextCoordinateTest, ConvertToDeviceSpace) {
+    const std::vector<CGPoint>& userSpacePoints = ::testing::get<1>(GetParam());
+    const std::vector<CGPoint>& deviceSpacePoints = ::testing::get<2>(GetParam());
```
Comment (aballway) [inline tests/unittests/CoreGraphics/CGContextTests.mm]: assert that both vectors have the same size

File: tests/unittests/CoreGraphics/CGContextTests.mm [hunk 7]
```diff
@@ -262,3 +274,148 @@ void cgContextPathCompare(NSArray* expected, NSArray* result) {
...
+    for (unsigned int i = 0; i < userSpacePoints.size(); ++i) {
```
Comment (aballway) [inline tests/unittests/CoreGraphics/CGContextTests.mm]: nit: size_t

File: tests/unittests/CoreGraphics/CGContextTests.mm (hunk 6 above)
Comment (DHowett-MSFT) [inline tests/unittests/CoreGraphics/CGContextTests.mm]: The test will immediately except if you do that; no point.

**Pull request merged**